source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs

```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
         readonly IAuthenticationConfigurationStore authenticationConfigurationStore;
         readonly DirectoryServicesUserCreationFromPrincipal supportsAutoUserCreationFromPrincipals;
         readonly IUserStore userStore;
+        readonly IIntegratedChallengeTracker integratedChallengeTracker;
 
         public IntegratedAuthenticationHandler(ILog log,
             IAuthCookieCreator tokenIssuer,
             IAuthenticationConfigurationStore authenticationConfigurationStore, 
             DirectoryServicesUserCreationFromPrincipal supportsAutoUserCreationFromPrincipals,
-            IUserStore userStore)
+            IUserStore userStore,
+            IIntegratedChallengeTracker integratedChallengeTracker)
         {
             this.log = log;
             this.tokenIssuer = tokenIssuer;
             this.authenticationConfigurationStore = authenticationConfigurationStore;
             this.supportsAutoUserCreationFromPrincipals = supportsAutoUserCreationFromPrincipals;
             this.userStore = userStore;
+            this.integratedChallengeTracker = integratedChallengeTracker;
         }
 
         public Task HandleRequest(HttpContext context)
         {
-            var result = TryAuthenticateRequest(context);
-            
-            var principal = result.User;
+            var state = GetLoginState(context);
 
-            // Decode the state object sent from the client (if there was one) so we can use those hints to build the most appropriate response
-            // If the state can't be interpreted, we will fall back to a safe-by-default behaviour, however:
-            //   1. Deep-links will not work because we don't know where the anonymous request originally wanted to go
-            //   2. Cookies may not have the Secure flag set properly when SSL Offloading is in play
-            LoginState state = null;
-            if (context.Request.Query.TryGetValue("state", out var stateString))
+            // Based on https://github.com/dotnet/runtime/blob/cf63e732fc6fb57c0ea97c1b4ca965acce46343a/src/libraries/System.Net.Security/src/System/Net/Security/NegotiateStreamPal.Windows.cs#L52
+            // we're being a bit cautious about using IsAuthenticated
+            if (string.IsNullOrWhiteSpace(context.User.Identity.Name))
             {
-                try
+                if (integratedChallengeTracker.IsConnectionKnown(context.Connection.Id))
```

Comment:
Consider a positive logic flow:
1. Handle the initial challenge case first - is this a new challenge?
2. Handle the failure case second - is this a failed challenge?